File: 2020/day03/puzzle3.py
file = open('puzzle3.in')
map = []
slopes = [
  [1, 1],
  [3, 1],
  [5, 1],
  [7, 1],
  [1, 2]
]

for line in file:
  map.append(line.strip())

treeproduct = 1;

for slope in slopes:
  right, down = slope
  trees = 0
  x = 0
  y = 0
  
  while y < len(map):
    if map[y][x] == '#': 
      trees += 1
    y = y + down
    x = (x + right) % (len(line))
  treeproduct *= trees

print(len(map))
print(treeproduct)

File: 2020/day03/puzzle3.ts
const input = await Deno.readTextFile("puzzle3.in");
const rows = input.split("\n");

const slopes = [
  { right: 1, down: 1 },
  { right: 3, down: 1 },
  { right: 5, down: 1 },
  { right: 7, down: 1 },
  { right: 1, down: 2 },
];

const isRowOk = (rowIndex: number, down: number) => rowIndex % down === 0;

const getColumnIndex = (rowIndex: number, right: number, rowLength: number) =>
  (rowIndex * right) % rowLength;

const getChar = (row: string, rowIndex: number, right: number) =>
  row[getColumnIndex(rowIndex, right, row.length)];

const isObstacle = (row: string, rowIndex: number, right: number) =>
  getChar(row, rowIndex, right) === "#";

const trees = slopes.map(
  ({ right, down }) =>
    rows
      .filter((row, rowIndex) => isRowOk(rowIndex, down))
      .filter((row, rowIndex) => isObstacle(row, rowIndex, right)).length
);

const product = trees.reduce((p, trees) => p * trees, 1);

console.log(rows.length);
console.log(product);
